In [1]:
import numpy as np
import pandas as pd
import sys
from bs4 import BeautifulSoup as BSoup
import requests

In [2]:
formLink = "http://isa.epfl.ch/imoniteur_ISAP/!GEDPUBLICREPORTS.filter?ww_i_reportModel=133685247"
showLink = "http://isa.epfl.ch/imoniteur_ISAP/!GEDPUBLICREPORTS.html?ww_x_GPS=-1&ww_i_reportModel=133685247"
r = requests.get(formLink)
soup = BSoup(r.text, "lxml")
selectFields = soup.find_all("select")
infoField=selectFields[0].find("option", text="Informatique")["value"]
infoFieldParam=selectFields[0]["name"]
bsSem1=selectFields[2].find("option", text="Bachelor semestre 1")["value"]
bsSem6=selectFields[2].find("option", text="Bachelor semestre 6")["value"]
semParam=selectFields[2]["name"]
allYears = [y["value"] for y in selectFields[1].find_all("option")[1:]]
yearParam = selectFields[1]["name"]
htmlradiobutton=soup.find("input", type="radio")

In [3]:
def showLinkGen(htmlP, htmlV, fieldP, fieldV, yearP, yearV, semP, semV):
  return "{}&{}={}&{}={}&{}={}&{}={}".format(showLink, htmlP, htmlV, fieldP, fieldV, yearP, yearV, semP, semV)
def get_bs_html(bsSem, year):
    link = showLinkGen(htmlradiobutton["name"], htmlradiobutton["value"], infoFieldParam, infoField, yearParam, 
                 year, semParam, bsSem)
    return requests.get(link)
def get_bs_dataframe(request_link):
  soup2 = BSoup(request_link.text, "lxml")
  elems=soup2.find_all("tr")[2:]
  titleinfo = soup2.find("font").text.split(', ')
  semester=titleinfo[2]
  semester=int(semester[(len(semester)-2):])
  year=titleinfo[1]
  all_data=[]
  for elem in elems:
    items=elem.find_all("td")
    gender = "M" if (items[0].text == "Monsieur") else "F"
    sciper = int(items[10].text)
    all_data.append({"Scipper": sciper, "Sex": gender, "Year": year, "Semester": semester})
  return pd.DataFrame(all_data)

In [4]:
def get_bs_alldata():
  all_data = []
  for bsSem in [bsSem1, bsSem6]:
    for year in allYears:
        all_data.append(get_bs_dataframe(get_bs_html(bsSem, year)))
  return pd.concat(all_data)
bachelor_data = get_bs_alldata()

In [5]:
bachelor_data

Scipper  Semester Sex       Year
0    235688         1   M  2016-2017
1    274015         1   M  2016-2017
2    268410         1   F  2016-2017
3    271464         1   M  2016-2017
4    274518         1   M  2016-2017
5    249613         1   M  2016-2017
6    262214         1   M  2016-2017
7    262239         1   M  2016-2017
8    257916         1   M  2016-2017
9    271508         1   F  2016-2017
10   269521         1   M  2016-2017
11   266684         1   M  2016-2017
12   258071         1   M  2016-2017
13   271608         1   F  2016-2017
14   204625         1   M  2016-2017
15   273765         1   M  2016-2017
16   266668         1   M  2016-2017
17   269502         1   M  2016-2017
18   274748         1   M  2016-2017
19   274746         1   M  2016-2017
20   270953         1   F  2016-2017
21   262179         1   M  2016-2017
22   274751         1   F  2016-2017
23   258330         1   F  2016-2017
24   260772         1   M  2016-2017
25   189560         1   M  2016-2017
26   270349         1   F  2016-2017
27   269905         1   M  2016-2017
28   275608         1   M  2016-2017
29   263800         1   M  2016-2017
..      ...       ...  ..        ...
8    153762         6   M  2007-2008
9    166548         6   M  2007-2008
10   166746         6   M  2007-2008
11   166748         6   M  2007-2008
12   166751         6   M  2007-2008
13   170414         6   M  2007-2008
14   166889         6   M  2007-2008
15   161091         6   F  2007-2008
16   173604         6   F  2007-2008
17   159998         6   F  2007-2008
18   170833         6   M  2007-2008
19   170823         6   M  2007-2008
20   170241         6   M  2007-2008
21   170242         6   M  2007-2008
22   165562         6   M  2007-2008
23   167007         6   M  2007-2008
24   166260         6   M  2007-2008
25   167313         6   M  2007-2008
26   171028         6   M  2007-2008
27   167386         6   M  2007-2008
28   159516         6   M  2007-2008
29   161127         6   M  2007-2008
30   170931         6   M  2007-2008
31   167045         6   M  2007-2008
32   161220         6   M  2007-2008
33   170975         6   M  2007-2008
34   167530         6   M  2007-2008
35   170265         6   M  2007-2008
36   167439         6   M  2007-2008
37   171042         6   M  2007-2008

[2384 rows x 4 columns]

In [6]:
data = bachelor_data.copy()

In [7]:
data.reset_index(None,drop=True,inplace=True)
data

Scipper  Semester Sex       Year
0      235688         1   M  2016-2017
1      274015         1   M  2016-2017
2      268410         1   F  2016-2017
3      271464         1   M  2016-2017
4      274518         1   M  2016-2017
5      249613         1   M  2016-2017
6      262214         1   M  2016-2017
7      262239         1   M  2016-2017
8      257916         1   M  2016-2017
9      271508         1   F  2016-2017
10     269521         1   M  2016-2017
11     266684         1   M  2016-2017
12     258071         1   M  2016-2017
13     271608         1   F  2016-2017
14     204625         1   M  2016-2017
15     273765         1   M  2016-2017
16     266668         1   M  2016-2017
17     269502         1   M  2016-2017
18     274748         1   M  2016-2017
19     274746         1   M  2016-2017
20     270953         1   F  2016-2017
21     262179         1   M  2016-2017
22     274751         1   F  2016-2017
23     258330         1   F  2016-2017
24     260772         1   M  2016-2017
25     189560         1   M  2016-2017
26     270349         1   F  2016-2017
27     269905         1   M  2016-2017
28     275608         1   M  2016-2017
29     263800         1   M  2016-2017
...       ...       ...  ..        ...
2354   153762         6   M  2007-2008
2355   166548         6   M  2007-2008
2356   166746         6   M  2007-2008
2357   166748         6   M  2007-2008
2358   166751         6   M  2007-2008
2359   170414         6   M  2007-2008
2360   166889         6   M  2007-2008
2361   161091         6   F  2007-2008
2362   173604         6   F  2007-2008
2363   159998         6   F  2007-2008
2364   170833         6   M  2007-2008
2365   170823         6   M  2007-2008
2366   170241         6   M  2007-2008
2367   170242         6   M  2007-2008
2368   165562         6   M  2007-2008
2369   167007         6   M  2007-2008
2370   166260         6   M  2007-2008
2371   167313         6   M  2007-2008
2372   171028         6   M  2007-2008
2373   167386         6   M  2007-2008
2374   159516         6   M  2007-2008
2375   161127         6   M  2007-2008
2376   170931         6   M  2007-2008
2377   167045         6   M  2007-2008
2378   161220         6   M  2007-2008
2379   170975         6   M  2007-2008
2380   167530         6   M  2007-2008
2381   170265         6   M  2007-2008
2382   167439         6   M  2007-2008
2383   171042         6   M  2007-2008

[2384 rows x 4 columns]

In [8]:
for i in range(data.shape[0]):
    if (data.loc[i,'Semester'] == 1):
        data.loc[i,'Year'] = data.loc[i,'Year'][0:4]
    else:
        data.loc[i,'Year'] = data.loc[i,'Year'][5:9]

data['Year'] = data['Year'].astype('int')
data.head()

Scipper  Semester Sex  Year
0   235688         1   M  2016
1   274015         1   M  2016
2   268410         1   F  2016
3   271464         1   M  2016
4   274518         1   M  2016

## keep only students who occur in sem1 and sem6

In [9]:
# only students in sem 1 and sem 6
idx_sem16 = (data.Scipper).isin(data[data.Semester == 6].Scipper) & (data.Scipper).isin(data[data.Semester == 1].Scipper)
data_sem16 = data[idx_sem16]
data_sem16.head()

Scipper  Semester Sex  Year
452   246671         1   M  2014
456   251759         1   M  2014
462   249498         1   M  2014
463   234551         1   M  2014
473   247328         1   M  2014

# compute stay time

In [10]:
scippers = data_sem16.Scipper.unique()
len(scippers)

397

In [11]:
tmp = pd.DataFrame({ 'Scipper' : scippers})
tmp['Staytime'] = np.nan
tmp = tmp.set_index(['Scipper'])
tmp.head()

Staytime
Scipper          
246671        NaN
251759        NaN
249498        NaN
234551        NaN
247328        NaN

In [12]:
data_sem16 = data_sem16.set_index(['Scipper'])

In [13]:
# students who graduated
for s in scippers:
    data_s = data_sem16.loc[s]
    nb = data_s.Year.max()-data_s.Year.min()
    nb = nb*12 - 6
    tmp.loc[s] = nb
tmp.reset_index(None,inplace=True)

In [14]:
tmp.head()

Scipper  Staytime
0   246671      30.0
1   251759      30.0
2   249498      30.0
3   234551      30.0
4   247328      30.0

In [15]:
data_sem16.reset_index(None,inplace=True)
data_sem16 = data_sem16.drop(['Year','Semester'],axis=1)
data_sem16 = data_sem16.drop_duplicates()
data_sem16

Scipper Sex
0     246671   M
1     251759   M
2     249498   M
3     234551   M
4     247328   M
5     251758   M
6     250300   M
7     249954   M
8     249996   M
9     239489   F
10    223892   M
11    246663   M
12    245511   M
13    250703   M
14    245512   M
15    247855   M
16    245782   M
17    247233   M
18    225452   M
19    250362   M
20    247680   M
21    246532   M
22    247891   M
23    228219   M
24    219665   F
25    227120   M
26    214939   M
27    236802   M
28    227370   M
29    226312   M
..       ...  ..
432   171619   M
433   179157   M
434   178711   M
435   178786   M
436   178656   M
437   178718   M
438   178706   F
439   181121   M
441   180570   M
442   178604   M
443   175190   M
444   181248   M
446   169731   M
447   178433   M
448   181298   M
449   181076   M
450   181115   M
451   180094   M
452   181017   M
453   179194   M
454   179988   F
455   180959   M
456   171195   M
457   178948   M
458   169795   M
459   178684   M
460   180241   M
461   175280   M
462   180854   M
463   178682   M

[397 rows x 2 columns]

In [16]:
data_sem16 = pd.merge(data_sem16, tmp, how='outer')
data_sem16

Scipper Sex  Staytime
0     246671   M      30.0
1     251759   M      30.0
2     249498   M      30.0
3     234551   M      30.0
4     247328   M      30.0
5     251758   M      30.0
6     250300   M      30.0
7     249954   M      30.0
8     249996   M      30.0
9     239489   F      30.0
10    223892   M      42.0
11    246663   M      30.0
12    245511   M      30.0
13    250703   M      30.0
14    245512   M      30.0
15    247855   M      30.0
16    245782   M      30.0
17    247233   M      30.0
18    225452   M      54.0
19    250362   M      30.0
20    247680   M      30.0
21    246532   M      30.0
22    247891   M      30.0
23    228219   M      30.0
24    219665   F      42.0
25    227120   M      30.0
26    214939   M      30.0
27    236802   M      30.0
28    227370   M      42.0
29    226312   M      42.0
..       ...  ..       ...
367   171619   M      30.0
368   179157   M      42.0
369   178711   M      42.0
370   178786   M      30.0
371   178656   M      30.0
372   178718   M      42.0
373   178706   F      30.0
374   181121   M      30.0
375   180570   M      30.0
376   178604   M      42.0
377   175190   M      30.0
378   181248   M      30.0
379   169731   M      42.0
380   178433   M      30.0
381   181298   M      30.0
382   181076   M      42.0
383   181115   M      30.0
384   180094   M      30.0
385   181017   M      30.0
386   179194   M      30.0
387   179988   F      30.0
388   180959   M      30.0
389   171195   M      30.0
390   178948   M      30.0
391   169795   M      42.0
392   178684   M      30.0
393   180241   M      30.0
394   175280   M      42.0
395   180854   M      30.0
396   178682   M      30.0

[397 rows x 3 columns]

# partitonning the data + comparing average

In [17]:
data_grouped = data_sem16.groupby('Sex')
data_grouped['Staytime'].mean()

Sex
F    33.724138
M    35.771739
Name: Staytime, dtype: float64

# Statistical tests

In [19]:
import scipy.stats as stats

## Divide Populations

In [20]:
data_F = data_sem16[data_sem16.Sex == 'F']
data_M = data_sem16[data_sem16.Sex == 'M']

## 1-sample T-Test

In a 1-sample T-Test, the null hypothesis assumes nothing interesting is going on between the variables we are testing. In this case, it means that there is no difference between each of the sub-populations and the whole population.

In [21]:
stats.ttest_1samp(data_M.Staytime,data_sem16.Staytime.mean())

Ttest_1sampResult(statistic=0.32081543725818445, pvalue=0.74853286145726683)

A p-value of 0.7485 means we'd expect to see data as extreme as our sample due to chance about 74.85% of the time if the null hypothesis was true. In this case, the p-value is higher than our significance level α (equal to 1-conf.level or 0.05) so we should not reject the null hypothesis.
#### => The stay-time average of males is statically significant.

In [22]:
stats.ttest_1samp(data_F.Staytime,data_sem16.Staytime.mean())

Ttest_1sampResult(statistic=-1.5732943251612512, pvalue=0.12688368345278089)

A p-value of 0.1268 means we'd expect to see data as extreme as our sample due to chance about only 12.68% of the time if the null hypothesis was true. 
In this case, the p-value is low than our significance level so we should reject the null hypothesis.
#### The stay-time average of females is not statically significant, since the female population is different.

## 2-sample T-Test

In a 2-sample T-Test, the null hypothesis states that the groups are the same.

In [23]:
stats.ttest_ind(a= data_M.Staytime, b = data_F.Staytime, equal_var = False)

Ttest_indResult(statistic=1.5831651359439409, pvalue=0.12191236829650401)

The test yields a p-value of 0.1219, which means there is a 12.19% chance we would see sample data this far apart if the two groups tested are actually identical.
#### => We conclude that the differance in the average of the stay-time between males and females is not statically significant.